irohad/ametsuchi/impl/mutable_storage_impl.cpp

```diff
@@ -71,17 +70,25 @@ namespace iroha {
                  block->height(),
                  block->hash().hex());
 
-      // TODO 09.04.2019 mboldyrev IR-440 add height check to predicate
-      auto block_applied = predicate(block, *peer_query_, top_hash_)
+      auto block_applied =
+          (not ledger_state_ or predicate(block, *ledger_state_.value()))
           and std::all_of(block->transactions().begin(),
                           block->transactions().end(),
                           execute_transaction);
       if (block_applied) {
         block_storage_->insert(block);
         block_index_->index(*block);
 
-        top_hash_ = block->hash();
-        top_height_ = block->height();
+        auto opt_ledger_peers = peer_query_->getLedgerPeers();
+        if (not opt_ledger_peers) {
+          log_->error("Failed to get ledger peers!");
+          return false;
+        }
+        assert(opt_ledger_peers);
+
+        // atomic?
```

Comment:
I don't think it's necessary.

Reply:
Ow, and I thought it should be synchronized. Summon @lebdron

Reply:
`MutableStorageImpl` is not expected to be used from different threads, it should not be necessary.